domain/blockdevice/state/state.go

```diff
@@ -0,0 +1,402 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/canonical/sqlair"
+	"github.com/google/uuid"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/blockdevice"
+)
+
+// State represents database interactions dealing with block devices.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new block device state
+// based on the input database factory method.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// BlockDevices returns the BlockDevices for the specified machine.
+func (st *State) BlockDevices(ctx context.Context, machine string) ([]blockdevice.BlockDevice, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var result []blockdevice.BlockDevice
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		result, err = loadBlockDevices(ctx, tx, machine)
+		return errors.Trace(err)
+	})
+	return result, errors.Trace(err)
+}
+
+func loadBlockDevices(ctx context.Context, tx *sqlair.TX, machine string) ([]blockdevice.BlockDevice, error) {
+	credQuery := `
+SELECT bd.* AS &BlockDevice.*,
+       bdl.* AS &DeviceLink.*,
+       fs_type.* AS &FilesystemType.*
+FROM   block_device bd
+       JOIN block_device_machine bdm ON bd.uuid = bdm.block_device_uuid
+       JOIN machine ON bdm.machine_uuid = machine.uuid
+       LEFT JOIN block_device_link_device bdl ON bd.uuid = bdl.block_device_uuid
+       LEFT JOIN filesystem_type fs_type ON bd.filesystem_type_id = fs_type.id
+WHERE  machine.machine_id = $M.machine_id
+`
+
+	types := []any{
+		BlockDevice{},
+		FilesystemType{},
+		DeviceLink{},
+		sqlair.M{},
+	}
+
+	stmt, err := sqlair.Prepare(credQuery, types...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		dbRows            BlockDevices
+		dbDeviceLinks     []DeviceLink
+		dbFilesystemTypes []FilesystemType
+	)
+	machineParam := sqlair.M{"machine_id": machine}
+	err = tx.Query(ctx, stmt, machineParam).GetAll(&dbRows, &dbDeviceLinks, &dbFilesystemTypes)
+	if err != nil {
+		return nil, errors.Annotatef(err, "loading block devices for machine %q", machine)
+	}
+	return dbRows.toBlockDevices(dbDeviceLinks, dbFilesystemTypes)
+}
+
+// GetMachineInfo is used by the block device service to look up
```

Comment:
This comment doesn't make any sense. Why are we documenting the users of the function? What if we add more users?

Just simple pre and post is enough and also expected errors. See comments around not found above.

Reply:
There's do reason to use this function except for the service - I was trying to signal that it should not be used except for this one place and it is only there out of begrudging necessity in order to allow the watcher to be set up. If we had a better agree pattern to set up the watchers - we do it differently in a couple of places to avoid this but the approach is under discussion - then we wouldn't need this.

Reply:
Yeah that is fine and makes sense. But its purpose isn't anything special. It exists and there is no way to reliably block it to one consumer. My argument is just document what it does and let the caller decide if it is fit for purpose. Just reads weird other wise.

Reply:
ok, not a hill i'll die on